src/Catalyst.Core.Lib.Tests/UnitTests/P2P/Protocols/PeerQueryTipTests.cs

```diff
@@ -0,0 +1,123 @@
+
+namespace Catalyst.Core.Lib.Tests.UnitTests.P2P.Protocols
+{
+    public sealed class PeerQueryTipTests : SelfAwareTestBase
+    {
+        private readonly IPeerQueryTipRequest _peerQueryTipRequest;
+        private readonly IPeerSettings _testSettings;
+        private readonly CancellationTokenProvider _cancellationProvider;
+
+        public PeerQueryTipTests(ITestOutputHelper output) : base(output)
+        {
+            var subbedPeerClient = Substitute.For<IPeerClient>();
+            _testSettings = PeerSettingsHelper.TestPeerSettings();
+            _cancellationProvider = new CancellationTokenProvider(TimeSpan.FromSeconds(10));
+            
+            _peerQueryTipRequest = new PeerQueryTipRequestRequest(
+                Substitute.For<ILogger>(),
+                subbedPeerClient,
+                _testSettings,
+                _cancellationProvider
+            );
+        }
+
+        [Fact]
+        public async Task Can_Query_Expected_Peer()
+        {
+            var recipientPeerId = PeerIdHelper.GetPeerId();
+            await _peerQueryTipRequest.QueryPeerTipAsync(recipientPeerId).ConfigureAwait(false);
+            var expectedDto = Substitute.For<IMessageDto<ProtocolMessage>>();
+            expectedDto.RecipientPeerIdentifier.Returns(recipientPeerId);
+            expectedDto.SenderPeerIdentifier.Returns(_testSettings.PeerId);
+            _peerQueryTipRequest.PeerClient.ReceivedWithAnyArgs(1).SendMessage(Arg.Is(expectedDto));
+        }
+
+        [Fact]
+        public async Task Can_Receive_Query_Response_On_Observer()
+        {
+            var recipientPeerId = PeerIdHelper.GetPeerId();
+            var tipQueryResponse = new PeerQueryTipResponse(PeerIdHelper.GetPeerId(), 
+                MultiHash.ComputeHash(ByteUtil.GenerateRandomByteArray(32))
+            );
+
+            _peerQueryTipRequest.QueryTipResponseMessageStreamer.OnNext(tipQueryResponse);
+            var response = await _peerQueryTipRequest.QueryPeerTipAsync(recipientPeerId).ConfigureAwait(false);
+            response.Should().BeTrue();
+        }
+
+        [Fact]
+        public async Task No_Response_Timeout_And_Returns_False()
+        {
+            var recipientPeerId = PeerIdHelper.GetPeerId();
+            _cancellationProvider.CancellationTokenSource.Cancel();
+            var response = await _peerQueryTipRequest.QueryPeerTipAsync(recipientPeerId).ConfigureAwait(false);
+            response.Should().BeFalse();
+        }
+
+        [Fact]
+        public async Task Exception_During_Query_Returns_Null()
+        {
+            var recipientPeerId = PeerIdHelper.GetPeerId();
+            _cancellationProvider.Dispose(); //do summet nasty to force exception
+            var response = await _peerQueryTipRequest.QueryPeerTipAsync(recipientPeerId).ConfigureAwait(false);
+            response.Should().BeFalse();   
+        }
+
+        // [Fact]
+        // public async Task Can_Dispose_Class()
+        // {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this commented out code.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4600656)